<a href="https://colab.research.google.com/github/gymnopedies7/tensorflow_tutorial/blob/master/tensorflow_tutorial_iris.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Library 불러오기
import tensorflow as tf
import pandas as pd

In [13]:
# Data 준비
파일경로 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
아이리스 = pd.read_csv(파일경로)
아이리스.head()
아이리스[['품종']].nunique()

품종    3
dtype: int64

In [14]:
# Data dummy 처리(원핫인코딩)

one_hot = pd.get_dummies(아이리스)
one_hot.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_setosa  품종_versicolor  품종_virginica
0   5.1  3.5    1.4   0.2          1              0             0
1   4.9  3.0    1.4   0.2          1              0             0
2   4.7  3.2    1.3   0.2          1              0             0
3   4.6  3.1    1.5   0.2          1              0             0
4   5.0  3.6    1.4   0.2          1              0             0

In [17]:
# 독립변수, 종속변수 구분
독립 = one_hot[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
종속 = one_hot[['품종_setosa', '품종_versicolor','품종_virginica']]

In [18]:
print(독립.shape, 종속.shape)

(150, 4) (150, 3)


In [19]:
# 모델 구조만들기

X = tf.keras.layers.Input(shape=[4])
Y = tf.keras.layers.Dense(3, activation = 'softmax')(X)
model = tf.keras.models.Model(X,Y)
model.compile(loss='categorical_crossentropy',
              metrics = 'accuracy')

In [22]:
model.fit(독립,종속, epochs=500, verbose=0)


In [24]:
model.fit(독립,종속, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2476 - accuracy: 0.9600
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2478 - accuracy: 0.9667
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2467 - accuracy: 0.9600
Epoch 4/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2466 - accuracy: 0.9667
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2462 - accuracy: 0.9600
Epoch 6/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2463 - accuracy: 0.9667
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2457 - accuracy: 0.9667
Epoch 8/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2454 - accuracy: 0.9667
Epoch 9/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2449 - accuracy: 0.9667
Epoch 10/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2446 - accuracy: 0.9667


In [26]:
print(model.predict(독립[:5]))
print(종속[:5])

[[9.9226499e-01 7.6812091e-03 5.3769756e-05]
 [9.7721577e-01 2.2571802e-02 2.1241703e-04]
 [9.8763788e-01 1.2242994e-02 1.1917685e-04]
 [9.7457719e-01 2.5119584e-02 3.0322064e-04]
 [9.9356925e-01 6.3839643e-03 4.6734483e-05]]
   품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0


In [27]:
print(model.get_weights())

[array([[ 0.15640467, -0.07459181, -0.81703967],
       [ 2.0659382 , -0.02807743, -0.32292947],
       [-2.9179645 , -0.42044136,  0.13669784],
       [-1.5749046 , -0.7434585 ,  1.8623216 ]], dtype=float32), array([ 0.8241325,  0.8072348, -0.637271 ], dtype=float32)]


Weight를 보면 

1열 0.156 , 2.065, -2.917, -1.574, (0.824)


이게 setosa 에 대한 가중치임. 그리고 마지막 0.824는 편향(bias)